src/privacy_accounting/optimizer/dp_optimizer.py

```diff
@@ -0,0 +1,100 @@
+from math import ceil
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+from src.privacy_accounting.analysis import QueryWithLedger
+from src.privacy_accounting.dp_query import GaussianDPQuery
+from . import WrapperOptimiser
+
+
+class DPOptimiser(WrapperOptimiser):
+
+    def __init__(self,
+                 optimiser,
+                 model,
+                 vector_loss,
+                 dp_sum_query,
+                 num_microbatches=None):
+
+        self.optimiser = optimiser
+        self.model = model
+        self.vector_loss = vector_loss
+        self.dp_sum_query = dp_sum_query
+        self.num_microbatches = num_microbatches
+
+        self._global_parameters = self.dp_sum_query.initial_global_state()
+
+    def fit_batch(self, x: torch.Tensor, y: torch.Tensor):
+
+        loss = self.vector_loss(self.model(x), y)
```

Comment:
yes - nn.module overrides the class __call__ function so that when doing that ^ the __call__ function is called, does some logic with hooks, calls the .forward function and the cleans up

Reply:
Go check nn.Module.__call__ for more details